djinn/djinn.1.in
.TH djinn 1 "@RELEASE_DATE@" "@PACKAGE_STRING@" "User Commands"
.SH NAME
djinn \- UNIX daemon wrapper for Java programs
.SH SYNOPSIS
\fBdjinn\fP [\fB-j\fP \fIjvm\fP] [[\fB-J\fP \fIarg\fP] ...]
[\fB-p\fP \fIclasspath\fP] [\fB-o\fP \fIlogfile\fP]
[\fB--help --version --verbose\fP] \fIclass\fP [\fIarg\fP ...]
.SH DESCRIPTION

\fBdjinn\fP is a simple C wrapper program that allows a Java program
to run as a daemon. It invokes the Java Virtual Machine on the named
\fIclass\fP in a subprocess that has no standard I/O streams and no
controlling terminal. Any arguments which follow the class name are
passed to the \fBmain()\fP function in the class. This subprocess is
created as the leader of its own process group, and its signal mask is
modified to ignore the \fBSIGHUP\fP signal.

The absolute path to the Java Virtual Machine may be specified on the
command line or through the \fBJAVA\fP environment variable, with the
command line switch taking precedence. If neither of these is
specified, \fBdjinn\fP attempts to locate an executable named "java" in the 
binary search path, and if successful, uses that as the JVM.

After spawning the JVM, if \fBdjinn\fP detects that the JVM
immediately exited with an error, it writes a message to that effect
to standard error; any messages from the JVM itself will be available
in the log file, if one has been specified.

.SH OPTIONS
.TP 5
.B --help
Display a command synopsis and exit.
.TP 5
.B --version
Display the program's version information and exit.
.TP 5
.B --verbose
Run in verbose mode.
.TP 5
.B -j \fIjvm\fP
Specify the absolute path to the Java VM executable. If this switch is
omitted, \fBdjinn\fP will check the \fBJAVA\fP environment variable,
and if that does not contain the path of an executable file, it will
attempt to find a program named ``java'' in the binary search path, and
use that as the JVM.
.TP 5
.B -J \fIarg\fP
Pass the argument \fIarg\fP to the Java Virtual
Machine. If several arguments need to be passed, this switch may be repeated.
The argument \fIarg\fP should be enclosed in quotes to prevent the
shell from interpreting it as a switch to \fBdjinn\fP itself.
If no occurrences of this switch appear in the argument list, \fBdjinn\fP
will check the \fBJAVA_ARGS\fP environment variable, and if it is
defined, will pass the arguments specified therein to the JVM.
.TP 5
.B -o \fIlogfile\fP
Specify an output file; the Java VM's standard output and standard
error streams will be redirected into \fIlogfile\fP. If this switch is 
omitted, the JVM's output is sent to /dev/null.
.TP 5
.B -p \fIclasspath\fP
Specify the \fIclasspath\fP for the Java Virtual Machine. If this switch is
omitted, the value of the \fBCLASSPATH\fP environment variable will be
used instead.
.SH EXAMPLES
.TP 5
The following example invokes the client JVM on the class \fBcom.foo.Test\fP, passing the arguments \fI10\fP and \fIHello\fP to its \fBmain()\fP function.

.B
djinn -J "-client" -p . com.foo.Test 10 Hello
.P
.SH ENVIRONMENT
.TP 15
.B JAVA
An absolute path to a Java Virtual Machine executable.
.TP 15
.B JAVA_ARGS
Any additional arguments to pass to the JVM.
.TP 15
.B CLASSPATH
The classpath for the JVM.
.TP 15
.B PATH
The binary search path for locating the JVM executable.
.SH NOTES
\fIJava\fP is a trademark of Sun Microsystems, Inc.
.SH AUTHOR
.PD 0
.TP 5
Mark Lindner <@PACKAGE_BUGREPORT@>


ChangeLog
2017-10-28  Mark Lindner  <markl@avalon>

	* stat/stat.c - fixed compiler warning
	* watcherd/* - removed
	* pfortune/* - removed
	* configure.ac - bumped version

2012-10-01  Mark Lindner  <markl@neuromancer>

	* aux-build/* - updated files
	* pkgenv/pkgenv.c - fixed compiler warnings

2011-06-26  Mark Lindner  <markl@neuromancer>

	* bin2c/* - added new utility
	* configure.ac, Makefile.am - added new bin2c utility, bump version

2008-04-23  Mark Lindner  <markl@neuromancer>

	* pfortune.1.in - fix typo in manpage

2008-04-21  Mark Lindner  <markl@neuromancer>

	* pkgenv.c - bugfix for shell handling

2008-04-20  Mark Lindner  <markl@neuromancer>

	* texi2ps, texi2ps.1.in - new script
	* pkgenv.c - bugfix: implement proper version string comparator
	* configure.ac - bump version

2007-12-10  Mark Lindner  <markl@neuromancer>

	* man2ps, man2ps.1.in, ps2pdf, ps2pdf.1.in, texi2pdf, texi2pdf.1.in,
	  man2pdf, man2pdf.1.in - modified scripts to accept multiple filename
	  arguments
	* <multiple> - rename 'tformat' utility to 'wrap'
	* pkgenv.c - fix memory crash bug

2007-12-08  Mark Lindner  <markl@neuromancer>

	* ranline.c - bugfix

2007-12-06  Mark Lindner  <markl@neuromancer>

	* ranline.c, ranline.1.in - add support for reading from stdin;
	  increase maximum supported line length to 1024

2007-11-19  Mark Lindner  <markl@neuromancer>

	* pkgenv.c, pkgenv.1.in - added -V (verbose) flag
	* basecvt.c, basecvt.1.in - added '?' command

2007-11-11  Mark Lindner  <markl@neuromancer>

	* pkgenv.c - improve error messages
	* pkgenv/automake_patch.txt - added

2007-09-16  Mark Lindner  <markl@neuromancer>

	* pkgenv.c, pkgenv.1.in - allow environment variable references in
	  variable substitution

2007-09-05  Mark Lindner  <markl@neuromancer>

	* pkgenv.c, pkgenv.1.in - more bugfixes & improvements

2007-09-01  Mark Lindner  <markl@neuromancer>

	* pkgenv.c - add special logic to handle old vs. new MANPATH dir;
	  don't add dirs that don't exist; bugfixes to 'del' command;
	  add DYLD_LIBRARY_PATH support for Mac OS X
	* *.1, *.1.in, configure.ac - generate manpages with correct version &
	  date in header
	* scripts/man2pdf - new script
	* Makefile.am - add 'pdf' target

2007-08-26  Mark Lindner  <markl@neuromancer>

	* <multiple> - code cleanup, bugfixes
	* pkgenv/, basecvt/ - new utilities added

2007-02-08  Mark Lindner  <markl@neuromancer>

	* scripts/Makefile.am - add manpages to dist
	* pfortune.c - fix segmentation fault

2006-04-17  Mark Lindner  <markl@neuromancer>

	* scripts/ - added useful shell scripts

2006-04-17  Mark Lindner  <markl@neuromancer>

	* dirstack.c, textlock.c, configure.ac - check for both ncurses.h
	  and ncurses/ncurses.h; fixes build error on OS X.

2006-04-15  Mark Lindner  <markl@neuromancer>

	* pfortune.c, pfortune.1 - partial rewrite, new -l switch
	* pfortune - added new databases (smileys, paul_graham, fortunes)

2006-04-13  Mark Lindner  <markl@neuromancer>

	* <multiple> - modified build files to locate libcfl using pkg-config
	* smileys.fdb - new fortune database added

2005-08-15  Mark Lindner  <markl@neuromancer>

	* dirstack.1, pascii.1 - manpage corrections
	* README - updated

2005-08-12  Mark Lindner  <markl@neuromancer>

	* <multiple> - Merged in 'djinn' project.
	* dirstack.c - Truncate paths that do not fit on one line.

2004-05-12  Mark Lindner  <markl@neuromancer>

	* bat.c - Added -8 switch.
	* <multiple> - Various minor code fixes.

2003-02-22  Mark Lindner  <markl@neuromancer>

	* <multiple> - Reorganized package so that there is only one
	  configure.ac (at the top level)
	* <multiple> - Modified all source code to use CFL library (which
	  replaces the pingutil library)
	* stat.c - don't exit with an error if getgrnam() or getpwent() fails,
	  but rather just display "?" for the user or group name
	* dirstack.c - fixed bug that created repaint glitch on delete,
	  added handling of ^L for screen refresh, and added -p switch.
	* ranline.c - use random() instead of rand() so that (on Solaris, at
	  least) we're not limited to a random number range of 32767.
	* <multiple> - updated manpages
	* pascii.c - handle DEL key and high-bit chars more gracefully
	* textlock.c - use C_system_get_hostname() to get hostname; the old way
	  of copying the results of uname() was non-portable, because SYS_NMLN
	  is not defined on some systems (namely OS X).

2003-01-02  Mark Lindner  <markl@neuromancer>

	* <multiple> - replaced `bool_t' with `p_bool_t' to fix compilation
	  errors when linked against the latest pingutil

2002-04-17  Mark Lindner  <mark_a_lindner@yahoo.com>

	* <multiple> - reorganized the entire source tree to use automake &
	  autoconf. Damn this was a lot of work (especially since I'd never
	  used automake before). Also made several minor fixes to make the
	  packages compile cleanly on Linux.

2002-04-14  Mark Lindner  <mark_a_lindner@yahoo.com>

	* dirstack - new utility

2001-10-14  Mark Lindner  <mark_a_lindner@yahoo.com>

	* <multiple> - minor modifications for pingutil-2.2
	* watcherd - new utility
	* pfortune - new utility

2001-09-14  Mark Lindner  <mark_a_lindner@yahoo.com>

	* <multiple> - slight modifications to use pingutil-2.1
	* ptree - renamed to dirtree to avoid confusion with Solaris ptree cmd
	* dirtree - fixed segmentation fault


djinn/djinn.c
/* ----------------------------------------------------------------------------
   djinn - UNIX daemon wrapper for Java programs
   Copyright (C) 1998-2011  Mark Lindner

   This file is part of misctools.

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License as
   published by the Free Software Foundation; either version 2 of
   the License, or (at your option) any later version.

   This software is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this software; if not, see
   <http://www.gnu.org/licenses/>.
   ----------------------------------------------------------------------------
*/

/* --- System Headers --- */

#include "config.h"

#include <getopt.h>
#include <stdio.h>
#ifdef HAVE_STDLIB_H
#include <stdlib.h>
#endif /* HAVE_STDLIB_H */
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif /* HAVE_UNISTD_H */
#include <stddef.h>
#include <string.h>
#ifdef HAVE_SYS_TYPES_H
#include <sys/types.h>
#endif /* HAVE_SYS_TYPES_H */
#ifdef HAVE_SYS_STAT_H
#include <sys/stat.h>
#endif /* HAVE_SYS_STAT_H */
#ifdef HAVE_FCNTL_H
#include <fcntl.h>
#endif /* HAVE_FCNTL_H */
#include <signal.h>
#include <sys/wait.h>
#include <unistd.h>

#include <cbase/cbase.h>

/* --- Macros --- */

#define ENV_JAVA "JAVA"
#define ENV_JAVA_ARGS "JAVA_ARGS"
#define DEVNULL "/dev/null"
#define VM_NAME "java"

#define USAGE "[-p classpath] [-j jvm] [[-J vmarg] ...]\n\t" \
    "[-o file] [--version --help --verbose] class [arg ...]\n"
#define HEADER "djinn v" VERSION " - Mark Lindner\n"

#define HELP_OPTION 0
#define VERSION_OPTION 1
#define VERBOSE_OPTION 2

static const struct option options[] = { { "help", 0, NULL, 0  },
                                         { "version", 0, NULL, 0 },
                                         { "verbose", 0, NULL, 0 },
                                         { NULL, 0, NULL, 0 } };

static int verbose = 0;

/* --- Functions --- */

int main(int argc, char **argv)
{
  char *javavm = NULL, *classpath = NULL, *class = NULL, *logfile = NULL,
    **q, **xargs;
  c_vector_t *jargs, *vmargs, *args;
  int c, i, option_index, fd = -1, errflag = 0, vmargflag = 0;
  pid_t pid;

  C_error_init(*argv);

  vmargs = C_vector_start(20);

  /* parse the command line */

  while((c = getopt_long(argc, argv, "+J:j:o:p:", options, &option_index))
        != EOF)
  {
    switch(c)
    {
      case 'J':
        C_vector_store(vmargs, optarg);
        vmargflag = 1;
        break;

      case 'p':
        classpath = optarg;
        break;

      case 'j':
        javavm = optarg;
        break;

      case 'o':
        logfile = optarg;
        break;

      case 0:
      {
        switch(option_index)
        {
          case HELP_OPTION:
            C_error_printf("%s\n", HEADER);
            C_error_usage(USAGE);
            exit(EXIT_SUCCESS);
            break;

          case VERSION_OPTION:
            C_error_printf("%s\n", HEADER);
            exit(EXIT_SUCCESS);
            break;

          case VERBOSE_OPTION:
            verbose = 1;
            break;

          default:
            errflag = 1;
            break;
        }
        break;
      }

      default:
        errflag = 1;
        break;
    }
  }

  /* catch illegal option errors */

  if(errflag || (argc <= optind))
  {
    C_error_printf("%s\n", HEADER);
    C_error_usage(USAGE);
    exit(EXIT_FAILURE);
  }

  /* parse out class file */

  class = argv[optind++];

  jargs = C_vector_start(20);

  for(i = optind; i < argc; i++)
    C_vector_store(jargs, argv[i]);

  /* check for optional java args; first check command line, and if no
     -J switches were specified, resort to the environment variable
     JAVA_ARGS
  */

  if(! vmargflag)
  {
    char *p;
    char *args = getenv(ENV_JAVA_ARGS);

    if(args)
    {
      for(p = strtok(args, "\t "); p; p = strtok(NULL, "\t "))
        C_vector_store(vmargs, p);
    }
  }

  /* find the Java VM; first check the command line, and if the -j switch
     wasn't specified, resort to the environment variable JAVA
  */

  if(!javavm)
    javavm = getenv(ENV_JAVA);

  /* see if VM is valid */

  if(javavm)
    if(access(javavm, X_OK))
    {
      C_error_printf("Cannot execute JVM: %s\n", javavm);
      exit(EXIT_FAILURE);
    }

  /* construct the complete argument list */

  args = C_vector_start(20);

  // java [-classpath CLASSPATH] [VM ARGS] CLASS [ARGS]

  C_vector_store(args, VM_NAME);

  if(classpath != NULL)
  {
    // add classpath switch if classpath specified

    C_vector_store(args, "-classpath");
    C_vector_store(args, classpath);
  }

  // add any other args to the list

  if((q = C_vector_end(vmargs, NULL)))
    for(; *q; q++)
      C_vector_store(args, *q);

  C_vector_store(args, class);

  if((q = C_vector_end(jargs, NULL)))
    for(; *q; q++)
      C_vector_store(args, *q);

  xargs = C_vector_end(args, NULL);

  if(verbose)
  {
    C_error_printf("Invoking:");
    for(q = xargs; *q; q++)
      fprintf(stderr, " %s", *q);
    fputc('\n', stderr);
  }

  /* open logfile, if one specified */

  if(!logfile)
    logfile = C_string_dup(DEVNULL);

  if((fd = open(logfile, O_RDWR | O_CREAT | O_TRUNC, S_IRUSR | S_IWUSR)) < 0)
  {
    C_error_printf("Unable to open logfile: %s\n", logfile);
    exit(EXIT_FAILURE);
  }
  else
  {
    if(verbose)
      C_error_printf("Using logfile: %s\n", logfile);
  }

  /* other miscellany */

  signal(SIGHUP, SIG_IGN);

  /* fork subprocess and run VM */

  pid = fork();
  switch(pid)
  {
    case 0:
    {
      /* child */

      (void)setsid();
      dup2(fd, STDOUT_FILENO);
      dup2(fd, STDERR_FILENO);

      if(! javavm)
        execvp(VM_NAME, xargs);
      else
        execvp(javavm, xargs);

      exit(EXIT_FAILURE);
    }

    case -1:
    {
      /* fork error */

      C_error_printf("Unable to fork subprocess.\n");
      exit(EXIT_FAILURE);
      break;
    }

    default:
    {
      int r;

      /* parent */

      if(verbose)
        C_error_printf("Daemon process started (pid: %i)\n", (int)pid);

      /* Wait 1 second and then try to collect the exit status from the
       * subprocess. If the JVM exited abnormally, we'll detect it and
       * notify the user.
       */

      sleep(1);
      if((waitpid(pid, &r, WNOHANG) == pid) && (r != 0))
      {
        C_error_printf("JVM exited abnormally with exit code %i\n",
                       WEXITSTATUS(r));
        exit(EXIT_FAILURE);
      }

      break;
    }
  }

  exit(EXIT_SUCCESS);
}

/* end of source file */
